File: java/src/org/freemedsoftware/json/Relay.java
/*
 * $Id: Relay.java 114 2009-05-05 14:48:03Z jeff $
 *
 * Authors:
 *      Adam Buchbinder <adam.buchbinder@gmail.com>
 *      Jeff Buchbinder <jeff@freemedsoftware.org>
 *
 * FreeMED Electronic Medical Record and Practice Management System
 * Copyright (C) 1999-2009 FreeMED Software Foundation
 *
 * This program is free software; you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation; either version 2 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program; if not, write to the Free Software
 * Foundation, Inc., 675 Mass Ave, Cambridge, MA 02139, USA.
 */

package org.freemedsoftware.json;

import java.io.BufferedReader;
import java.io.IOException;
import java.io.InputStream;
import java.io.InputStreamReader;
import java.io.PrintStream;
import java.io.UnsupportedEncodingException;
import java.net.CookieHandler;
import java.net.HttpURLConnection;
import java.net.URI;
import java.net.URL;
import java.net.URLEncoder;
import java.util.ArrayList;
import java.util.Collections;
import java.util.HashMap;
import java.util.List;
import java.util.Map;
import java.util.Properties;
import java.util.logging.Logger;

import javax.net.ssl.HttpsURLConnection;

import net.sf.sojo.interchange.Serializer;
import net.sf.sojo.interchange.json.JsonSerializer;

import org.apache.log4j.PropertyConfigurator;

/**
 * Relay class for communicating with FreeMED 0.9.x+.
 */
public class Relay {
	/**
	 * Java 1.5 has the CookieHandler API, but not the default CookieManager
	 * implementation. *Retarded*. Copied from
	 * http://java.sun.com/j2se/1.5.0/docs/guide/net/http-cookie.html
	 */
	class MyCookieHandler extends CookieHandler {
		/**
		 * Internal cookiejar to hold cookies.
		 */
		private List<String> cookieJar = new ArrayList<String>();

		public Map<String, List<String>> get(URI uri,
				Map<String, List<String>> rqstHdrs) throws IOException {
			// the cookies will be included in request
			Map<String, List<String>> map = new HashMap<String, List<String>>();
			List<String> l = cookieJar;
			map.put("Cookie", l);
			return Collections.unmodifiableMap(map);
		}

		/**
		 * Add all response headers to the internal cookiejar.
		 */
		@SuppressWarnings("unchecked")
		public void put(URI uri, Map responseHeaders) throws IOException {
			cookieJar.addAll((List) responseHeaders.get("Set-Cookie"));
		}
	}

	protected static Logger log = Logger.getLogger("Relay");

	private boolean DEBUG = true;

	private boolean loggedIn = false;

	private String host = "", path = "", user = "", pass = "", protocol = "";

	private int port = 80;

	public boolean isLoggedIn() {
		return loggedIn;
	}

	public String getHost() {
		return host;
	}

	public void setHost(String host) {
		this.host = host;
		loggedIn = false;
	}

	public String getPath() {
		return path;
	}

	public void setPath(String path) {
		this.path = path;
		loggedIn = false;
	}

	public String getUser() {
		return user;
	}

	public void setUser(String user) {
		this.user = user;
		loggedIn = false;
	}

	private String getPass() {
		return pass;
	}

	public void setPass(String pass) {
		this.pass = pass;
		loggedIn = false;
	}

	public int getPort() {
		return port;
	}

	public void setPort(int port) {
		this.port = port;
		loggedIn = false;
	}

	public String getProtocol() {
		return protocol;
	}

	public void setProtocol(String protocol) {
		this.protocol = protocol;
		loggedIn = false;
	}

	public Relay() {
		loggedIn = false;

		// Initialize logger if it
		try {
			String log4jfile = System.getProperty("log4j.properties");
			// System.out.println("log4j-properties: " + log4jfile);
			if (log4jfile != null) {
				String propertiesFilename = log4jfile;
				PropertyConfigurator.configure(propertiesFilename);
				log.info("Logger initialized.");
			} else {
				// System.out.println("Error setting up logger.");
			}
		} catch (Exception ex) {
			System.out.println("Couldn't set up log4j : " + ex);
		}

		// TODO make sure we really need to do this
		CookieHandler.setDefault(new MyCookieHandler());
	}

	/**
	 * Load FreeMED server connection information from a Java properties file.
	 * 
	 * @param fileName
	 *            Fully qualified path to properties file.
	 * @throws IOException
	 */
	public void loadProperties(String fileName) throws IOException {
		Properties props = new Properties();
		InputStream inputStream = ClassLoader
				.getSystemResourceAsStream(fileName);
		if (inputStream == null)
			throw new IOException("Couldn't open " + fileName + ".");
		props.load(inputStream);
		inputStream.close();
		protocol = props.getProperty("freemed.protocol");
		host = props.getProperty("freemed.host");
		path = props.getProperty("freemed.path");
		port = Integer.parseInt(props.getProperty("freemed.port"));
		user = props.getProperty("freemed.user");
		pass = props.getProperty("freemed.pass");
	}

	/**
	 * Log in.
	 * 
	 * @returns true if the user is logged in; false if it failed.
	 */
	public boolean logIn() {
		if (isLoggedIn())
			return true;
		try {
			if (call("org.freemedsoftware.public.Login.Validate",
					new Object[] { getUser(), getPass() }).equals("true")) {
				return true;
			} else {
				return false;
			}
		} catch (UnsupportedEncodingException uee) {
			log.warning("Unsupported encoding");
			return false;
		}
	}

	/**
	 * Call a named JSON method on the FreeMED server.
	 * 
	 * @param method
	 *            Fully qualified method name (example:
	 *            "org.freemedsoftware.public.Login.Validate")
	 * @param params
	 *            Array of objects to serialize into the JSON call as
	 *            parameters.
	 * @return
	 * @throws UnsupportedEncodingException
	 *             If JSON encoding isn't supported, throws this exception.
	 */
	public String call(String method, Object[] params)
			throws UnsupportedEncodingException {
		if (!isLoggedIn() && !method.matches(".*public.*")) {
			if (DEBUG) {
				System.out.println("Must be logged in first.");
			}
			log.warning("Must be logged in first.");
			return null;
		}

		HttpURLConnection conn = null;
		try {
			URL url = new URL(protocol, host, port, path + method);
			if (protocol.compareToIgnoreCase("http") == 0) {
				conn = (HttpURLConnection) url.openConnection();
			} else if (protocol.compareToIgnoreCase("https") == 0) {
				conn = (HttpsURLConnection) url.openConnection();
			} else {
				throw new Exception("Invalid protocol " + protocol
						+ " specified");
			}
			conn.setRequestMethod("POST");
		} catch (IOException e) {
			System.out.println("Failed to open connection: " + e);
			return null;
		} catch (Exception e) {
			System.out.println("Failed to create connection: " + e);
		}
		conn.setRequestProperty("Content-type",
				"application/x-www-form-urlencoded");
		conn.setUseCaches(false);
		conn.setDoOutput(true);
		conn.setDoInput(true);
		// construct array from params
		StringBuffer sb = new StringBuffer();
		int count = 0;
		for (Object p : params) {
			if (count > 0)
				sb.append("&");
			Serializer serializer = new JsonSerializer();
			String json = serializer.serialize(p).toString();
			sb.append("param" + count + "=" + URLEncoder.encode(json, "UTF-8"));
			count++;
		}

		StringBuffer result = new StringBuffer();
		try {
			PrintStream os = new PrintStream(conn.getOutputStream());
			os.print(sb.toString());
			os.flush();
			os.close();

			BufferedReader reader = new BufferedReader(new InputStreamReader(
					conn.getInputStream()));
			String line;
			while ((line = reader.readLine()) != null) {
				result.append(line);
			}
			reader.close();

			conn.disconnect();
		} catch (IOException e) {
			log.warning("IO failure " + e.toString());
			System.out.println("IO failure: " + e);
			e.printStackTrace();
			return null;
		}

		// TODO JSON-decode?
		return result.toString();
	}

	public static void main(String[] argv) {
		Relay r = new Relay();
		try {
			r.loadProperties("freemed.properties");
		} catch (IOException e) {
			System.out
					.println("Couldn't find freemed.properties in classpath.");
			System.exit(1);
		}
		if (r.logIn()) {
			System.out.println("Login successful.");
		} else {
			System.out.println("Login failed.");
		}
		/*
		 * String result = r.call("org.freemedsoftware.public.Login.Validate",
		 * new Object[] { r.getUser(), r.getPass() });
		 * System.out.println("Result = "+result);
		 */
	}
}

File: csharp/Relay.cs
/*
 * $Id: Relay.cs 44 2007-02-06 13:14:22Z jeff $
 *
 * Authors:
 * 	Jeff Buchbinder <jeff@freemedsoftware.org>
 *
 * FreeMED Electronic Medical Record and Practice Management System
 * Copyright (C) 1999-2013 FreeMED Software Foundation
 *
 * This program is free software; you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation; either version 2 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program; if not, write to the Free Software
 * Foundation, Inc., 675 Mass Ave, Cambridge, MA 02139, USA.
 *
 */

using System;
using System.IO;
using System.Net;
using System.Reflection;
using System.Text;
using System.Web;

using Newtonsoft.Json;

namespace org.freemedsoftware.FreeMED {

public class Relay {
	private string _uri;
	private string _username;
	private string _password;
	private bool _debug;
	private CookieCollection _cookies;

	public Relay ( ) {
		_cookies = null;
		_uri = null;
		_username = null;
		_password = null;
		_debug = false;
	}

	public bool Debug {
		get { return _debug; }
		set { _debug = value; }
	}
	
	public void SetParameters ( string s_uri, string s_username, string s_password ) {
		_uri = s_uri;
		_username = s_username;
		_password = s_password;
	} // end method SetParameters

	public bool Login ( ) {
		if ( (_uri.Length < 5) || (_username.Length < 3) ) {
			Console.WriteLine("Need to call SetParameters()");
			return false;
		}
		return (bool) Call( 
			"org.freemedsoftware.public.Login.Validate",
			new object[] {
				_username,
				_password
			}
		);
	} // end method Login

	public object Call ( string method, object[] parameters ) {
		return Call ( method, parameters, null );
	}
	public object Call ( string method, object[] parameters, Type type ) {
		string outputValue = CallRaw( method, parameters, type );
		object deserializedObject;
		if ( type == null ) {
			deserializedObject = (object) JavaScriptConvert.DeserializeObject(outputValue);
		} else {
			deserializedObject = JavaScriptConvert.DeserializeObject(outputValue, type);
		}
		if ( _debug ) { Console.WriteLine( "DEBUG: deserialized = {0}", deserializedObject.ToString() ); }
		return deserializedObject;
	}
	public string CallRaw ( string method, object[] parameters, Type type ) {
		string fullUri = String.Format( "{0}/relay.php/json/{1}", _uri, method );
		if ( _debug ) { Console.WriteLine( "DEBUG: POST to {0} with {1}", fullUri, parameters.ToString() ); }
		HttpWebRequest webRequest = (HttpWebRequest) HttpWebRequest.Create( fullUri );
		webRequest.ContentType = "application/x-www-form-urlencoded";
		webRequest.Method = "POST";
		webRequest.CookieContainer = new CookieContainer();
		if (_cookies != null) {
			webRequest.CookieContainer.Add( _cookies );
		}

		// Form parameters
		string formedParameters = "";
		for (int i=0; i<parameters.Length; i++) {
			Type objectType = parameters[i].GetType();
			if ( _debug ) { Console.WriteLine("DEBUG: Processing parameter {0} type = {1}", i, objectType.ToString() ); }
			if ( formedParameters.Length > 0 ) { formedParameters += "&"; }
			switch (objectType.ToString()) {
				case "System.Int32":
				formedParameters += "param" + i.ToString() + "=" + parameters[i].ToString();
				break;

				case "System.String":
				formedParameters += "param" + i.ToString() + "=" + parameters[i];
				break;

				default:
				if ( _debug ) { Console.WriteLine( "DEBUG: serializing non-understood parameter" ); }
				// Check for ToJson() method
				try {
					MethodInfo[] mi = objectType.GetMethods();
					bool found = false;
					foreach (MethodInfo m in mi) {
						if ( m.Name == "ToJson" ) {
							if ( _debug ) { Console.WriteLine( "DEBUG: found ToJson() method" ); }
							found = true;
						}
					}
					if (found) {
						MethodInfo ToJson = objectType.GetMethod ("ToJson");
						formedParameters += "param" + i.ToString() + "=" + System.Web.HttpUtility.UrlEncode( (string) ToJson.Invoke (parameters[i], null) );
					} else {
						throw new Exception( "parameter" );
					}
				} catch (Exception) {
					formedParameters += "param" + i.ToString() + "=" + (string) JavaScriptConvert.SerializeObject( parameters[i] );
				}
				break;
			}
		}

		if ( _debug ) { Console.WriteLine( "DEBUG: formedParameters = '{0}'", formedParameters ); }
		byte[] bytes = Encoding.ASCII.GetBytes( formedParameters );
		Stream os = null;
		try {
			webRequest.ContentLength = bytes.Length;
			os = webRequest.GetRequestStream( );
			os.Write( bytes, 0, bytes.Length );
		} catch ( WebException ex ) {
			Console.WriteLine( "ERROR: Call() caught exception for {0} ({1})", fullUri, ex.Message );
			return null;
		} finally {
			if (os != null) { os.Close( ); }
		}

		try {
			HttpWebResponse webResponse = (HttpWebResponse) webRequest.GetResponse( );
			if ( webResponse == null ) { return null; }
			StreamReader sr = new StreamReader( webResponse.GetResponseStream( ) );
			string outputValue = sr.ReadToEnd( ).Trim( );
			if ( _debug ) { Console.WriteLine( "DEBUG: Returned {0}", outputValue ); }

			// Pass cookie value back
			if (webResponse.Cookies.Count > 0) {
				_cookies = webResponse.Cookies;
			}
			if ( _debug ) { Console.WriteLine( "DEBUG: Finished importing new cookies, if any" ); }

			try {
				if ( _debug ) { Console.WriteLine( "DEBUG: About to examine outputValue" ); }
				if ( ( outputValue == null ) || ( outputValue == "" ) ) {
					if ( _debug ) { Console.WriteLine( "DEBUG: NULL output value found" ); }
					return "false";
				}
				if ( _debug ) { Console.WriteLine( "DEBUG: About to return outputValue" ); }
				return outputValue;
			} catch (NullReferenceException iEx) {
				Console.WriteLine( "ERROR: Caught exception for {0}", iEx.Message );
				return "";
			}
		} catch ( WebException ex ) {
			Console.WriteLine( "ERROR: Caught exception for {0} ({1})", fullUri, ex.Message );
		} 

		return null;
	} // end Call

} // end class

} // end namespace

